com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingMain.raytrace

```diff
@@ -66,12 +77,46 @@ void RayGen()
     jitteredPixelCoord.x += GetSample(currentPixelCoord, _RaytracingSampleIndex, 40);
     jitteredPixelCoord.y += GetSample(currentPixelCoord, _RaytracingSampleIndex, 41);
 
-    // Compute the ray direction, from those coordinates
-    float3 directionWS = -normalize(mul(jitteredPixelCoord, (float3x3)_PixelCoordToViewDirWS));
+    float3 directionWS;
+    float3 cameraPosWS;
+
+    float apertureRadius = _PathTracedDoFConstants.x;
+    if (apertureRadius == 0.0)
+    {
+        // Compute the ray direction from those coordinates (fast path for zero aperture)
+        directionWS = -normalize(mul(jitteredPixelCoord, (float3x3)_PixelCoordToViewDirWS));
+        cameraPosWS = _WorldSpaceCameraPos;
+    }
+    else
+    {
+        // Compute the ray origin and direction for a lens with non-zero aperture
+
+        // Apply the inverse viewport transform to go from vieport coordinates to NDC
```

Comment:
Am a bit confused, if we need to go from screen position to NDC why do we need a matrix mul? Both live on the image plane so it is just a multiplication + bias. Unless there is some nomenclature override here. 

Also typo vieport  instead of viewport

Reply:
Looking at the code you are building a matrix to build a world space direction? ComputePixelCoordToWorldSpaceViewDirectionMatrix

Reply:
I cannot use directly the ComputePixelCoordToWorldSpaceViewDirectionMatrix here, because the viewMatrix changes per pixel. So I'm using this function with an identity viewmatrix on the C# side, to get the matrix that properly scales the coordinates (including the lens shift parameters, fov etc) and then I generate the proper viewMatrix per pixel.

Looking at the ComputePixelCoordToWorldSpaceViewDirectionMatrix you will also see why I prefer to use the matrix form (the lens shift parameters make the matrix form more convenient, although if I decompose it to MADs I might save some instructions at the expense of code readability).

Reply:
It is still one MAD only per pixel given that most of your matrix is still a 2d scale matrix mul by a 2d translation matrix. If (using names from that function and pseudocoding here) 


Scale = float2( aspectRatio, -1.0f) * 2 * screenSize.zw * tanHalfVertFoV
Bias = float2(aspectRatio, 1) * float2(1 - 2*lensShift.xy) *  tanHalfVertFoV

I think should be equivalent (I might have messed up some signs here, it is back of the napkin math :D But here for the rationale, not the code itself). 

I find a bit more confusing to see in code something that mentions "transform to world space direction" to be used to move to NDC? I am ok if you prefer to keep it in matrix form (granted that is _far_ more computation per pixel), but then I think the call to ComputePixelCoordToWorldSpaceViewDirectionMatrix in ComputeInverseViewportMatrix should be commented more.